File: walletfreak_app/src/api/endpoints/notifications.ts
import apiClient from '../client';

export interface Notification {
  id: string;
  uid: string;
  type: 'blog_published' | 'benefit_expiration' | 'system';
  title: string;
  body: string;
  read: boolean;
  created_at: any;
  metadata: Record<string, any>;
  action_url: string | null;
  action_route: string | null;
}

export const getNotifications = async (limit = 20, cursor?: string) => {
  const params: Record<string, any> = { limit };
  if (cursor) params.cursor = cursor;
  const { data } = await apiClient.get('/notifications/', { params });
  return data as { notifications: Notification[] };
};

export const getUnreadCount = async () => {
  const { data } = await apiClient.get('/notifications/unread-count/');
  return data as { count: number };
};

export const markRead = async (id: string) => {
  const { data } = await apiClient.post(`/notifications/${id}/read/`);
  return data;
};

export const markAllRead = async () => {
  const { data } = await apiClient.post('/notifications/read-all/');
  return data;
};

export const deleteNotification = async (id: string) => {
  const { data } = await apiClient.delete(`/notifications/${id}/`);
  return data;
};

File: walletfreak/api/routers/notifications.py
from ninja import Router
from django.http import JsonResponse
from core.services import db
from api.auth_middleware import BearerAuth
import json

router = Router(tags=["notifications"], auth=BearerAuth())


@router.get("/")
def list_notifications(request):
    """List user's notifications (paginated via ?limit=20&cursor=doc_id)."""
    uid = request.auth
    try:
        limit = int(request.GET.get('limit', 20))
        limit = min(limit, 50)
        cursor = request.GET.get('cursor')
        notifications = db.get_user_notifications(uid, limit=limit, start_after=cursor)
        return {"notifications": notifications}
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


@router.get("/unread-count/")
def unread_count(request):
    """Return the number of unread notifications."""
    uid = request.auth
    try:
        count = db.get_unread_count(uid)
        return {"count": count}
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


@router.post("/{notification_id}/read/")
def mark_read(request, notification_id: str):
    """Mark a single notification as read."""
    uid = request.auth
    try:
        ok = db.mark_notification_read(uid, notification_id)
        if not ok:
            return JsonResponse({"error": "Not found"}, status=404)
        return {"success": True}
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


@router.post("/read-all/")
def mark_all_read(request):
    """Mark all notifications as read."""
    uid = request.auth
    try:
        count = db.mark_all_notifications_read(uid)
        return {"success": True, "count": count}
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


@router.delete("/{notification_id}/")
def delete_notification(request, notification_id: str):
    """Delete a single notification."""
    uid = request.auth
    try:
        ok = db.delete_notification(uid, notification_id)
        if not ok:
            return JsonResponse({"error": "Not found"}, status=404)
        return {"success": True}
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)
